File: src/api/inventory.ts
// Mock API functions for Inventory module
// In a real application, these would make HTTP requests to your backend

// Types
interface Product {
  id: string;
  name: string;
  category: 'consumables' | 'equipment' | 'medications' | 'supplies';
  batchNumber: string;
  vendor: string;
  costPrice: number;
  sellingPrice?: number;
  currentStock: number;
  minStockLevel: number;
  maxStockLevel: number;
  unit: 'pieces' | 'ml' | 'grams' | 'boxes' | 'vials';
  expiryDate?: string;
  manufacturingDate?: string;
  location: string;
  description?: string;
  isActive: boolean;
  createdAt: string;
  updatedAt: string;
  lastUsed?: string;
  autoDeductEnabled: boolean;
  treatmentTypes: string[]; // Which treatments this product is used for
}

interface InventoryLog {
  id: string;
  productId: string;
  productName: string;
  type: 'stock-in' | 'stock-out' | 'adjustment' | 'expired' | 'auto-deduct';
  quantity: number;
  previousStock: number;
  newStock: number;
  reason: string;
  treatmentId?: string;
  patientName?: string;
  performedBy: string;
  createdAt: string;
  notes?: string;
}

interface InventoryStats {
  totalProducts: number;
  lowStockAlerts: number;
  expiringSoon: number;
  autoDeductToday: number;
  totalValue: number;
  categoriesCount: number;
}

interface StockAdjustment {
  productId: string;
  quantity: number;
  type: 'add' | 'remove';
  reason: string;
  notes?: string;
}

interface InventoryFilters {
  category?: string;
  vendor?: string;
  expiryFrom?: string;
  expiryTo?: string;
  stockLevel?: 'all' | 'low' | 'normal' | 'high';
  search?: string;
}

// Mock data storage (using localStorage for persistence)
const STORAGE_KEYS = {
  PRODUCTS: 'hospverse_inventory_products',
  INVENTORY_LOGS: 'hospverse_inventory_logs'
};

// Helper function to get data from localStorage
const getStorageData = <T>(key: string, defaultValue: T[] = []): T[] => {
  try {
    const data = localStorage.getItem(key);
    return data ? JSON.parse(data) : defaultValue;
  } catch (error) {
    console.error('Error reading from localStorage:', error);
    return defaultValue;
  }
};

// Helper function to save data to localStorage
const setStorageData = <T>(key: string, data: T[]): void => {
  try {
    localStorage.setItem(key, JSON.stringify(data));
  } catch (error) {
    console.error('Error saving to localStorage:', error);
  }
};

// Generate unique ID
const generateId = (): string => Date.now().toString(36) + Math.random().toString(36).substr(2);

// Mock vendors
const VENDORS = [
  'MedSupply Corp',
  'Healthcare Solutions Ltd',
  'BioMed Distributors',
  'Pharma Plus',
  'Medical Equipment Co',
  'Aesthetic Supplies Inc'
];

// Mock treatment types
const TREATMENT_TYPES = [
  'Laser Hair Removal',
  'PRP Treatment',
  'Chemical Peel',
  'Microneedling',
  'Botox Injection',
  'Dermal Fillers',
  'Acne Treatment',
  'Consultation'
];

// Initialize with some mock data if localStorage is empty
const initializeMockData = (): void => {
  if (!localStorage.getItem(STORAGE_KEYS.PRODUCTS)) {
    const mockProducts: Product[] = [
      {
        id: '1',
        name: 'Laser Gel',
        category: 'consumables',
        batchNumber: 'LG2024001',
        vendor: 'MedSupply Corp',
        costPrice: 25,
        sellingPrice: 35,
        currentStock: 15,
        minStockLevel: 10,
        maxStockLevel: 50,
        unit: 'ml',
        expiryDate: new Date(Date.now() + 180 * 24 * 60 * 60 * 1000).toISOString().split('T')[0],
        manufacturingDate: new Date(Date.now() - 30 * 24 * 60 * 60 * 1000).toISOString().split('T')[0],
        location: 'Storage Room A',
        description: 'Cooling gel for laser treatments',
        isActive: true,
        createdAt: new Date(Date.now() - 30 * 24 * 60 * 60 * 1000).toISOString(),
        updatedAt: new Date(Date.now() - 2 * 60 * 60 * 1000).toISOString(),
        lastUsed: new Date(Date.now() - 2 * 60 * 60 * 1000).toISOString(),
        autoDeductEnabled: true,
        treatmentTypes: ['Laser Hair Removal']
      },
      {
        id: '2',
        name: 'PRP Tubes',
        category: 'supplies',
        batchNumber: 'PRP2024002',
        vendor: 'BioMed Distributors',
        costPrice: 15,
        currentStock: 8,
        minStockLevel: 15,
        maxStockLevel: 100,
        unit: 'pieces',
        expiryDate: new Date(Date.now() + 365 * 24 * 60 * 60 * 1000).toISOString().split('T')[0],
        manufacturingDate: new Date(Date.now() - 60 * 24 * 60 * 60 * 1000).toISOString().split('T')[0],
        location: 'Refrigerator B',
        description: 'Sterile tubes for PRP preparation',
        isActive: true,
        createdAt: new Date(Date.now() - 60 * 24 * 60 * 60 * 1000).toISOString(),
        updatedAt: new Date(Date.now() - 24 * 60 * 60 * 1000).toISOString(),
        lastUsed: new Date(Date.now() - 24 * 60 * 60 * 1000).toISOString(),
        autoDeductEnabled: true,
        treatmentTypes: ['PRP Treatment']
      },
      {
        id: '3',
        name: 'Chemical Peel Solution',
        category: 'medications',
        batchNumber: 'CPS2024003',
        vendor: 'Pharma Plus',
        costPrice: 80,
        sellingPrice: 120,
        currentStock: 25,
        minStockLevel: 5,
        maxStockLevel: 30,
        unit: 'ml',
        expiryDate: new Date(Date.now() + 90 * 24 * 60 * 60 * 1000).toISOString().split('T')[0],
        manufacturingDate: new Date(Date.now() - 90 * 24 * 60 * 60 * 1000).toISOString().split('T')[0],
        location: 'Chemical Storage',
        description: 'TCA 20% solution for chemical peels',
        isActive: true,
        createdAt: new Date(Date.now() - 90 * 24 * 60 * 60 * 1000).toISOString(),
        updatedAt: new Date(Date.now() - 3 * 24 * 60 * 60 * 1000).toISOString(),
        lastUsed: new Date(Date.now() - 3 * 24 * 60 * 60 * 1000).toISOString(),
        autoDeductEnabled: true,
        treatmentTypes: ['Chemical Peel']
      },
      {
        id: '4',
        name: 'Microneedling Cartridges',
        category: 'supplies',
        batchNumber: 'MNC2024004',
        vendor: 'Aesthetic Supplies Inc',
        costPrice: 12,
        currentStock: 45,
        minStockLevel: 20,
        maxStockLevel: 100,
        unit: 'pieces',
        expiryDate: new Date(Date.now() + 730 * 24 * 60 * 60 * 1000).toISOString().split('T')[0],
        manufacturingDate: new Date(Date.now() - 30 * 24 * 60 * 60 * 1000).toISOString().split('T')[0],
        location: 'Treatment Room 1',
        description: 'Sterile cartridges for microneedling device',
        isActive: true,
        createdAt: new Date(Date.now() - 30 * 24 * 60 * 60 * 1000).toISOString(),
        updatedAt: new Date(Date.now() - 5 * 24 * 60 * 60 * 1000).toISOString(),
        lastUsed: new Date(Date.now() - 5 * 24 * 60 * 60 * 1000).toISOString(),
        autoDeductEnabled: true,
        treatmentTypes: ['Microneedling']
      },
      {
        id: '5',
        name: 'Botox Vials',
        category: 'medications',
        batchNumber: 'BTX2024005',
        vendor: 'Pharma Plus',
        costPrice: 300,
        sellingPrice: 450,
        currentStock: 3,
        minStockLevel: 5,
        maxStockLevel: 20,
        unit: 'vials',
        expiryDate: new Date(Date.now() + 60 * 24 * 60 * 60 * 1000).toISOString().split('T')[0],
        manufacturingDate: new Date(Date.now() - 120 * 24 * 60 * 60 * 1000).toISOString().split('T')[0],
        location: 'Refrigerator A',
        description: 'Botulinum toxin type A, 100 units',
        isActive: true,
        createdAt: new Date(Date.now() - 120 * 24 * 60 * 60 * 1000).toISOString(),
        updatedAt: new Date(Date.now() - 7 * 24 * 60 * 60 * 1000).toISOString(),
        lastUsed: new Date(Date.now() - 7 * 24 * 60 * 60 * 1000).toISOString(),
        autoDeductEnabled: true,
        treatmentTypes: ['Botox Injection']
      }
    ];
    setStorageData(STORAGE_KEYS.PRODUCTS, mockProducts);
  }

  if (!localStorage.getItem(STORAGE_KEYS.INVENTORY_LOGS)) {
    const mockLogs: InventoryLog[] = [
      {
        id: '1',
        productId: '1',
        productName: 'Laser Gel',
        type: 'auto-deduct',
        quantity: 5,
        previousStock: 20,
        newStock: 15,
        reason: 'Used in laser hair removal treatment',
        treatmentId: 'T001',
        patientName: 'John Doe',
        performedBy: 'Sarah Wilson',
        createdAt: new Date(Date.now() - 2 * 60 * 60 * 1000).toISOString(),
        notes: 'Automatic deduction for laser treatment'
      },
      {
        id: '2',
        productId: '2',
        productName: 'PRP Tubes',
        type: 'auto-deduct',
        quantity: 2,
        previousStock: 10,
        newStock: 8,
        reason: 'Used in PRP treatment',
        treatmentId: 'T002',
        patientName: 'Jane Smith',
        performedBy: 'Mike Johnson',
        createdAt: new Date(Date.now() - 24 * 60 * 60 * 1000).toISOString(),
        notes: 'Automatic deduction for PRP treatment'
      },
      {
        id: '3',
        productId: '3',
        productName: 'Chemical Peel Solution',
        type: 'stock-in',
        quantity: 50,
        previousStock: 0,
        newStock: 50,
        reason: 'New stock received',
        performedBy: 'Admin',
        createdAt: new Date(Date.now() - 90 * 24 * 60 * 60 * 1000).toISOString(),
        notes: 'Initial stock entry'
      }
    ];
    setStorageData(STORAGE_KEYS.INVENTORY_LOGS, mockLogs);
  }
};

// Initialize mock data
initializeMockData();

// API Functions
export const getProducts = async (filters: InventoryFilters = {}): Promise<Product[]> => {
  return new Promise((resolve) => {
    setTimeout(() => {
      let products = getStorageData<Product>(STORAGE_KEYS.PRODUCTS);
      
      // Apply filters
      if (filters.category && filters.category !== 'all') {
        products = products.filter(product => product.category === filters.category);
      }
      
      if (filters.vendor && filters.vendor !== 'all') {
        products = products.filter(product => product.vendor === filters.vendor);
      }
      
      if (filters.stockLevel && filters.stockLevel !== 'all') {
        products = products.filter(product => {
          const stockRatio = product.currentStock / product.maxStockLevel;
          switch (filters.stockLevel) {
            case 'low':
              return product.currentStock <= product.minStockLevel;
            case 'normal':
              return product.currentStock > product.minStockLevel && stockRatio <= 0.8;
            case 'high':
              return stockRatio > 0.8;
            default:
              return true;
          }
        });
      }
      
      if (filters.expiryFrom && filters.expiryTo) {
        const fromDate = new Date(filters.expiryFrom);
        const toDate = new Date(filters.expiryTo);
        products = products.filter(product => {
          if (!product.expiryDate) return false;
          const expiryDate = new Date(product.expiryDate);
          return expiryDate >= fromDate && expiryDate <= toDate;
        });
      }
      
      if (filters.search) {
        const searchTerm = filters.search.toLowerCase();
        products = products.filter(product => 
          product.name.toLowerCase().includes(searchTerm) ||
          product.batchNumber.toLowerCase().includes(searchTerm) ||
          product.vendor.toLowerCase().includes(searchTerm)
        );
      }
      
      // Sort by name
      products.sort((a, b) => a.name.localeCompare(b.name));
      
      resolve(products);
    }, 300);
  });
};

export const getProductDetails = async (id: string): Promise<Product> => {
  return new Promise((resolve, reject) => {
    setTimeout(() => {
      const products = getStorageData<Product>(STORAGE_KEYS.PRODUCTS);
      const product = products.find(p => p.id === id);
      
      if (product) {
        resolve(product);
      } else {
        reject(new Error('Product not found'));
      }
    }, 200);
  });
};

export const addStock = async (productId: string, quantity: number, notes?: string): Promise<Product> => {
  return new Promise((resolve, reject) => {
    setTimeout(() => {
      const products = getStorageData<Product>(STORAGE_KEYS.PRODUCTS);
      const productIndex = products.findIndex(p => p.id === productId);
      
      if (productIndex === -1) {
        reject(new Error('Product not found'));
        return;
      }
      
      const product = products[productIndex];
      const previousStock = product.currentStock;
      const newStock = previousStock + quantity;
      
      // Update product
      products[productIndex] = {
        ...product,
        currentStock: newStock,
        updatedAt: new Date().toISOString()
      };
      
      // Add inventory log
      const logs = getStorageData<InventoryLog>(STORAGE_KEYS.INVENTORY_LOGS);
      const newLog: InventoryLog = {
        id: generateId(),
        productId,
        productName: product.name,
        type: 'stock-in',
        quantity,
        previousStock,
        newStock,
        reason: 'Stock added',
        performedBy: 'Current User', // In real app, get from auth context
        createdAt: new Date().toISOString(),
        notes
      };
      logs.unshift(newLog);
      
      setStorageData(STORAGE_KEYS.PRODUCTS, products);
      setStorageData(STORAGE_KEYS.INVENTORY_LOGS, logs);
      
      resolve(products[productIndex]);
    }, 500);
  });
};

export const deductProduct = async (productId: string, quantity: number, reason: string, treatmentId?: string, patientName?: string): Promise<Product> => {
  return new Promise((resolve, reject) => {
    setTimeout(() => {
      const products = getStorageData<Product>(STORAGE_KEYS.PRODUCTS);
      const productIndex = products.findIndex(p => p.id === productId);
      
      if (productIndex === -1) {
        reject(new Error('Product not found'));
        return;
      }
      
      const product = products[productIndex];
      
      if (product.currentStock < quantity) {
        reject(new Error('Insufficient stock'));
        return;
      }
      
      const previousStock = product.currentStock;
      const newStock = previousStock - quantity;
      
      // Update product
      products[productIndex] = {
        ...product,
        currentStock: newStock,
        lastUsed: new Date().toISOString(),
        updatedAt: new Date().toISOString()
      };
      
      // Add inventory log
      const logs = getStorageData<InventoryLog>(STORAGE_KEYS.INVENTORY_LOGS);
      const newLog: InventoryLog = {
        id: generateId(),
        productId,
        productName: product.name,
        type: 'auto-deduct',
        quantity,
        previousStock,
        newStock,
        reason,
        treatmentId,
        patientName,
        performedBy: 'System', // Auto-deduction
        createdAt: new Date().toISOString()
      };
      logs.unshift(newLog);
      
      setStorageData(STORAGE_KEYS.PRODUCTS, products);
      setStorageData(STORAGE_KEYS.INVENTORY_LOGS, logs);
      
      resolve(products[productIndex]);
    }, 400);
  });
};

export const adjustStock = async (adjustment: StockAdjustment): Promise<Product> => {
  return new Promise((resolve, reject) => {
    setTimeout(() => {
      const products = getStorageData<Product>(STORAGE_KEYS.PRODUCTS);
      const productIndex = products.findIndex(p => p.id === adjustment.productId);
      
      if (productIndex === -1) {
        reject(new Error('Product not found'));
        return;
      }
      
      const product = products[productIndex];
      const previousStock = product.currentStock;
      const quantity = adjustment.type === 'add' ? adjustment.quantity : -adjustment.quantity;
      const newStock = previousStock + quantity;
      
      if (newStock < 0) {
        reject(new Error('Stock cannot be negative'));
        return;
      }
      
      // Update product
      products[productIndex] = {
        ...product,
        currentStock: newStock,
        updatedAt: new Date().toISOString()
      };
      
      // Add inventory log
      const logs = getStorageData<InventoryLog>(STORAGE_KEYS.INVENTORY_LOGS);
      const newLog: InventoryLog = {
        id: generateId(),
        productId: adjustment.productId,
        productName: product.name,
        type: 'adjustment',
        quantity: Math.abs(quantity),
        previousStock,
        newStock,
        reason: adjustment.reason,
        performedBy: 'Current User', // In real app, get from auth context
        createdAt: new Date().toISOString(),
        notes: adjustment.notes
      };
      logs.unshift(newLog);
      
      setStorageData(STORAGE_KEYS.PRODUCTS, products);
      setStorageData(STORAGE_KEYS.INVENTORY_LOGS, logs);
      
      resolve(products[productIndex]);
    }, 500);
  });
};

export const getInventoryLogs = async (productId?: string): Promise<InventoryLog[]> => {
  return new Promise((resolve) => {
    setTimeout(() => {
      let logs = getStorageData<InventoryLog>(STORAGE_KEYS.INVENTORY_LOGS);
      
      if (productId) {
        logs = logs.filter(log => log.productId === productId);
      }
      
      // Sort by creation date (newest first)
      logs.sort((a, b) => new Date(b.createdAt).getTime() - new Date(a.createdAt).getTime());
      
      resolve(logs);
    }, 300);
  });
};

export const getInventoryStats = async (): Promise<InventoryStats> => {
  return new Promise((resolve) => {
    setTimeout(() => {
      const products = getStorageData<Product>(STORAGE_KEYS.PRODUCTS);
      const logs = getStorageData<InventoryLog>(STORAGE_KEYS.INVENTORY_LOGS);
      
      const totalProducts = products.filter(p => p.isActive).length;
      
      const lowStockAlerts = products.filter(p => 
        p.isActive && p.currentStock <= p.minStockLevel
      ).length;
      
      // Products expiring in next 30 days
      const thirtyDaysFromNow = new Date(Date.now() + 30 * 24 * 60 * 60 * 1000);
      const expiringSoon = products.filter(p => 
        p.isActive && p.expiryDate && new Date(p.expiryDate) <= thirtyDaysFromNow
      ).length;
      
      // Auto-deductions today
      const today = new Date().toISOString().split('T')[0];
      const autoDeductToday = logs.filter(log => 
        log.type === 'auto-deduct' && log.createdAt.split('T')[0] === today
      ).length;
      
      const totalValue = products
        .filter(p => p.isActive)
        .reduce((sum, p) => sum + (p.currentStock * p.costPrice), 0);
      
      const categoriesCount = new Set(products.filter(p => p.isActive).map(p => p.category)).size;
      
      resolve({
        totalProducts,
        lowStockAlerts,
        expiringSoon,
        autoDeductToday,
        totalValue,
        categoriesCount
      });
    }, 300);
  });
};

export const getVendors = async (): Promise<string[]> => {
  return new Promise((resolve) => {
    setTimeout(() => {
      resolve(VENDORS);
    }, 100);
  });
};

export const getTreatmentTypes = async (): Promise<string[]> => {
  return new Promise((resolve) => {
    setTimeout(() => {
      resolve(TREATMENT_TYPES);
    }, 100);
  });
};

// Export types for use in components
export type { Product, InventoryLog, InventoryStats, StockAdjustment, InventoryFilters };

File: api/inventory.py
from flask import Blueprint, request, jsonify, g
from .extensions import supabase
import uuid
from datetime import datetime

inventory_bp = Blueprint("inventory", __name__)

# Middleware to check if module is enabled
def check_module_access():
    if not g.tenant_id:
        return jsonify({"error": "Tenant not found"}), 404
        
    if "inventory" not in g.modules:
        return jsonify({"error": "Inventory module is not enabled for this tenant"}), 403
        
    return None

# Get all products
@inventory_bp.route("/inventory/products", methods=["GET"])
def get_products():
    # Check module access
    module_error = check_module_access()
    if module_error:
        return module_error
        
    try:
        # Apply filters if provided
        query = supabase.table("products").select("*").eq("client_id", g.tenant_id)
        
        category = request.args.get("category")
        if category and category != "all":
            query = query.eq("category", category)
            
        stock_level = request.args.get("stockLevel")
        if stock_level and stock_level != "all":
            if stock_level == "low":
                query = query.filter("current_stock", "lte", "min_stock_level")
            elif stock_level == "normal":
                query = query.filter("current_stock", "gt", "min_stock_level").filter("current_stock", "lte", "max_stock_level * 0.8")
            elif stock_level == "high":
                query = query.filter("current_stock", "gt", "max_stock_level * 0.8")
                
        expiry_from = request.args.get("expiryFrom")
        expiry_to = request.args.get("expiryTo")
        
        if expiry_from and expiry_to:
            query = query.gte("expiry_date", expiry_from).lte("expiry_date", expiry_to)
            
        search = request.args.get("search")
        if search:
            query = query.or_(f"name.ilike.%{search}%,batch_number.ilike.%{search}%,vendor.ilike.%{search}%")
            
        # Order by name
        query = query.order("name")
        
        result = query.execute()
        
        return jsonify(result.data), 200
    except Exception as e:
        return jsonify({"error": str(e)}), 500

# Get product details
@inventory_bp.route("/inventory/products/<product_id>", methods=["GET"])
def get_product_details(product_id):
    # Check module access
    module_error = check_module_access()
    if module_error:
        return module_error
        
    try:
        result = supabase.table("products").select("*").eq("id", product_id).eq("client_id", g.tenant_id).single().execute()
        
        if not result.data:
            return jsonify({"error": "Product not found"}), 404
            
        return jsonify(result.data), 200
    except Exception as e:
        return jsonify({"error": str(e)}), 500

# Add stock
@inventory_bp.route("/inventory/products/<product_id>/add-stock", methods=["POST"])
def add_stock(product_id):
    # Check module access
    module_error = check_module_access()
    if module_error:
        return module_error
        
    try:
        quantity = request.json.get("quantity")
        notes = request.json.get("notes")
        
        if not quantity or int(quantity) <= 0:
            return jsonify({"error": "Valid quantity is required"}), 400
            
        # Get current product
        product = supabase.table("products").select("*").eq("id", product_id).eq("client_id", g.tenant_id).single().execute()
        
        if not product.data:
            return jsonify({"error": "Product not found"}), 404
            
        # Calculate new stock
        previous_stock = product.data["current_stock"]
        new_stock = previous_stock + int(quantity)
        
        # Update product
        product_result = supabase.table("products").update({
            "current_stock": new_stock,
            "updated_at": datetime.now().isoformat()
        }).eq("id", product_id).execute()
        
        # Add inventory log
        log_id = str(uuid.uuid4())
        log = {
            "id": log_id,
            "client_id": g.tenant_id,
            "product_id": product_id,
            "product_name": product.data["name"],
            "type": "stock-in",
            "quantity": int(quantity),
            "previous_stock": previous_stock,
            "new_stock": new_stock,
            "reason": "Stock added",
            "performed_by": "Current User",  # In a real app, get from auth context
            "created_at": datetime.now().isoformat(),
            "notes": notes
        }
        
        supabase.table("inventory_logs").insert(log).execute()
        
        return jsonify(product_result.data[0]), 200
    except Exception as e:
        return jsonify({"error": str(e)}), 500

# Deduct product
@inventory_bp.route("/inventory/products/<product_id>/deduct", methods=["POST"])
def deduct_product(product_id):
    # Check module access
    module_error = check_module_access()
    if module_error:
        return module_error
        
    try:
        quantity = request.json.get("quantity")
        reason = request.json.get("reason")
        treatment_id = request.json.get("treatmentId")
        patient_name = request.json.get("patientName")
        
        if not quantity or int(quantity) <= 0:
            return jsonify({"error": "Valid quantity is required"}), 400
            
        if not reason:
            return jsonify({"error": "Reason is required"}), 400
            
        # Get current product
        product = supabase.table("products").select("*").eq("id", product_id).eq("client_id", g.tenant_id).single().execute()
        
        if not product.data:
            return jsonify({"error": "Product not found"}), 404
            
        # Check if sufficient stock
        previous_stock = product.data["current_stock"]
        quantity_int = int(quantity)
        
        if previous_stock < quantity_int:
            return jsonify({"error": "Insufficient stock"}), 400
            
        # Calculate new stock
        new_stock = previous_stock - quantity_int
        
        # Update product
        now = datetime.now().isoformat()
        product_result = supabase.table("products").update({
            "current_stock": new_stock,
            "last_used": now,
            "updated_at": now
        }).eq("id", product_id).execute()
        
        # Add inventory log
        log_id = str(uuid.uuid4())
        log = {
            "id": log_id,
            "client_id": g.tenant_id,
            "product_id": product_id,
            "product_name": product.data["name"],
            "type": "auto-deduct",
            "quantity": quantity_int,
            "previous_stock": previous_stock,
            "new_stock": new_stock,
            "reason": reason,
            "treatment_id": treatment_id,
            "patient_name": patient_name,
            "performed_by": "System",
            "created_at": now
        }
        
        supabase.table("inventory_logs").insert(log).execute()
        
        return jsonify(product_result.data[0]), 200
    except Exception as e:
        return jsonify({"error": str(e)}), 500

# Adjust stock
@inventory_bp.route("/inventory/products/adjust", methods=["POST"])
def adjust_stock():
    # Check module access
    module_error = check_module_access()
    if module_error:
        return module_error
        
    try:
        adjustment = request.json
        
        product_id = adjustment.get("productId")
        quantity = adjustment.get("quantity")
        adjustment_type = adjustment.get("type")
        reason = adjustment.get("reason")
        notes = adjustment.get("notes")
        
        if not product_id or not quantity or int(quantity) <= 0:
            return jsonify({"error": "Product ID and valid quantity are required"}), 400
            
        if not adjustment_type or adjustment_type not in ["add", "remove"]:
            return jsonify({"error": "Valid adjustment type (add/remove) is required"}), 400
            
        if not reason:
            return jsonify({"error": "Reason is required"}), 400
            
        # Get current product
        product = supabase.table("products").select("*").eq("id", product_id).eq("client_id", g.tenant_id).single().execute()
        
        if not product.data:
            return jsonify({"error": "Product not found"}), 404
            
        # Calculate new stock
        previous_stock = product.data["current_stock"]
        quantity_int = int(quantity)
        
        if adjustment_type == "add":
            new_stock = previous_stock + quantity_int
        else:  # remove
            if previous_stock < quantity_int:
                return jsonify({"error": "Stock cannot be negative"}), 400
            new_stock = previous_stock - quantity_int
            
        # Update product
        product_result = supabase.table("products").update({
            "current_stock": new_stock,
            "updated_at": datetime.now().isoformat()
        }).eq("id", product_id).execute()
        
        # Add inventory log
        log_id = str(uuid.uuid4())
        log = {
            "id": log_id,
            "client_id": g.tenant_id,
            "product_id": product_id,
            "product_name": product.data["name"],
            "type": "adjustment",
            "quantity": quantity_int,
            "previous_stock": previous_stock,
            "new_stock": new_stock,
            "reason": reason,
            "performed_by": "Current User",  # In a real app, get from auth context
            "created_at": datetime.now().isoformat(),
            "notes": notes
        }
        
        supabase.table("inventory_logs").insert(log).execute()
        
        return jsonify(product_result.data[0]), 200
    except Exception as e:
        return jsonify({"error": str(e)}), 500

# Get inventory logs
@inventory_bp.route("/inventory/logs", methods=["GET"])
def get_inventory_logs():
    # Check module access
    module_error = check_module_access()
    if module_error:
        return module_error
        
    try:
        # Apply filters if provided
        query = supabase.table("inventory_logs").select("*").eq("client_id", g.tenant_id)
        
        product_id = request.args.get("productId")
        if product_id:
            query = query.eq("product_id", product_id)
            
        # Order by creation date (newest first)
        query = query.order("created_at", desc=True)
        
        result = query.execute()
        
        return jsonify(result.data), 200
    except Exception as e:
        return jsonify({"error": str(e)}), 500

# Get inventory stats
@inventory_bp.route("/inventory/stats", methods=["GET"])
def get_inventory_stats():
    # Check module access
    module_error = check_module_access()
    if module_error:
        return module_error
        
    try:
        # Get products
        products_query = supabase.table("products").select("*").eq("client_id", g.tenant_id).execute()
        products = products_query.data
        
        # Get logs
        logs_query = supabase.table("inventory_logs").select("*").eq("client_id", g.tenant_id).execute()
        logs = logs_query.data
        
        # Calculate stats
        total_products = len([p for p in products if p["is_active"]])
        
        low_stock_alerts = len([
            p for p in products 
            if p["is_active"] and p["current_stock"] <= p["min_stock_level"]
        ])
        
        # Products expiring in next 30 days
        thirty_days_from_now = (datetime.now() + timedelta(days=30)).strftime("%Y-%m-%d")
        expiring_soon = len([
            p for p in products 
            if p["is_active"] and p.get("expiry_date") and p["expiry_date"] <= thirty_days_from_now
        ])
        
        # Auto-deductions today
        today = datetime.now().strftime("%Y-%m-%d")
        auto_deduct_today = len([
            log for log in logs 
            if log["type"] == "auto-deduct" and log["created_at"].startswith(today)
        ])
        
        # Calculate total value
        total_value = sum([
            p["current_stock"] * p["cost_price"] 
            for p in products if p["is_active"]
        ])
        
        # Count unique categories
        categories = set([p["category"] for p in products if p["is_active"]])
        categories_count = len(categories)
        
        stats = {
            "totalProducts": total_products,
            "lowStockAlerts": low_stock_alerts,
            "expiringSoon": expiring_soon,
            "autoDeductToday": auto_deduct_today,
            "totalValue": total_value,
            "categoriesCount": categories_count
        }
        
        return jsonify(stats), 200
    except Exception as e:
        return jsonify({"error": str(e)}), 500

# Get vendors
@inventory_bp.route("/inventory/vendors", methods=["GET"])
def get_vendors():
    # Check module access
    module_error = check_module_access()
    if module_error:
        return module_error
        
    try:
        # In a real app, this would fetch from the database
        # For this demo, we'll return mock data
        vendors = [
            "MedSupply Corp",
            "Healthcare Solutions Ltd",
            "BioMed Distributors",
            "Pharma Plus",
            "Medical Equipment Co",
            "Aesthetic Supplies Inc"
        ]
        
        return jsonify(vendors), 200
    except Exception as e:
        return jsonify({"error": str(e)}), 500

# Get treatment types
@inventory_bp.route("/inventory/treatment-types", methods=["GET"])
def get_treatment_types():
    # Check module access
    module_error = check_module_access()
    if module_error:
        return module_error
        
    try:
        # In a real app, this would fetch from the database
        # For this demo, we'll return mock data
        treatment_types = [
            "Laser Hair Removal",
            "PRP Treatment",
            "Chemical Peel",
            "Microneedling",
            "Botox Injection",
            "Dermal Fillers",
            "Acne Treatment",
            "Consultation"
        ]
        
        return jsonify(treatment_types), 200
    except Exception as e:
        return jsonify({"error": str(e)}), 500